airflow/contrib/hooks/gcp_dataflow_hook.py

```diff
@@ -300,3 +720,79 @@ def _start_template_dataflow(self, name, variables, parameters,
         _DataflowJob(self.get_conn(), variables['project'], name, variables['region'],
                      self.poll_sleep, num_retries=self.num_retries).wait_for_done()
         return response
+
+    def _add_execution_date_to_job_name(self, job_name, context):
+        """Adds execution date to the dataflow job name as suffix"""
+        self.log.info(
+            f"Suffixing the job name {job_name} with the execution time as 'append_job_name' "
+            f"is set to False"
+        )
+        new_job_name = f"{job_name}t{context['execution_date'].strftime('%Y-%m-%d-%H-%M')}"
+        self.log.info(f"New job name: {new_job_name}")
+        return new_job_name
+
+    @staticmethod
+    def _validate_params(cancel_existing_job, append_job_name):
+        # cancel_existing_job should only be set to True if append_job_name is set to False
+        if cancel_existing_job and append_job_name:
+            raise AirflowException(
+                "'cancel_existing_job' can only be used when 'append_job_name' is set to False"
+            )
+
+    def _track_or_cancel_existing_dataflow_job(self, job_name, variables, cancel_existing_job):
+        """Cancels any already running dataflow job with the same name"""
+        try:
+            project_id = variables['project']
+            location = variables['region']
+            dataflow_controller = _DataflowJobsController(
+                dataflow=self.get_conn(),
+                project_number=project_id,
+                name=job_name,
+                job_id=None,
+                location=location,
+                poll_sleep=self.poll_sleep,
+                num_retries=self.num_retries,
+                drain_pipeline=False,
+                cancel_timeout=5*60,
+                wait_until_finished=None
+            )
+
+            if cancel_existing_job:
+                self.log.info(f"Cancelling any already running Dataflow job with the name {job_name}")
+                dataflow_controller.cancel()
+                return False, None
+
+            else:
+                self.log.info(f"Tracking any already running Dataflow job with the name {job_name}")
+                jobs = dataflow_controller.get_jobs()
+                if not jobs:
+                    self.log.info("No existing jobs to track")
+                    return False, None
+
+                running_jobs = [
+                    job for job in jobs if job["currentState"] not in DataflowJobStatus.TERMINAL_STATES
+                ]
+
+                try:
+                    running_job_id = [job["id"] for job in running_jobs if job["name"] == job_name][0]
```

Comment:
should this be "job["name"].startswith(job_name)"? As we have given the option of appending the random string or time in the job name, it won't match the job name.
Am I missing something?

Reply:
We are leaving the original open source logic intact here. We are not tracking/cancelling jobs if random hash is added. Also, we cannot do "startswith" here since multiple dag runs can be active at the same time and they might interfere with each other then.